packages/lib/gpcircuits/src/util.ts

```diff
@@ -1,13 +1,89 @@
-import path from "path";
-import { CircuitArtifactPaths, CircuitDesc } from "./types";
+import * as fastfile from "fastfile";
+import { CircuitArtifactPaths, CircuitDesc, CircuitSignal } from "./types";
 
-export function artifactPaths(
+export type VerificationKey = object;
+
+/**
+ * Loads a verification key from a JSON file.
+ *
+ * @param vkeyPath path to load artifacts, which could be a URL (in browser)
+ *   or a file path (in server or utests).
+ * @returns the verification key as an object
+ * @throws if file access or JSON parsing fails
+ */
+export async function loadVerificationKey(
+  vkeyPath: string
+): Promise<VerificationKey> {
+  // This method of binary loading files using fastfile is the same as what is
+  // used in snarkjs to load pkey and wasm artifacts.  It works for local file
+  // paths (in Node) as well as URLs (in browser).
+  // The string decoding and JSON parsing is specific to our use of vkeys, which
+  // snarkjs assumes are already in memory via import.
+  let fd: fastfile.FastFile | undefined = undefined;
+  try {
+    fd = await fastfile.readExisting(vkeyPath);
+    const bytes = await fd.read(fd.totalSize);
+    return JSON.parse(Buffer.from(bytes).toString("utf8"));
+  } finally {
+    if (fd !== undefined) {
+      await fd.close();
+    }
+  }
+}
+
+/**
+ * Determines the right path for loading circuit artifacts for a given
+ * circuit.
+ *
+ * @param root root path to load artifacts, which could be a URL (in browser)
+ *   or a file path (in server or utests).
+ * @param cd description of the GPC circuit
+ * @returns collection of artifact paths
+ */
+export function gpcArtifactPaths(
   root: string,
   cd: CircuitDesc
 ): CircuitArtifactPaths {
+  if (!root.endsWith("/")) {
+    root = root + "/";
+  }
```

Comment:
why not `path.join`? i've found concatenating urls/paths to be a major footgun throughout this project so i've been trying to use exclusively `path.join` and `urljoin` (an npm package) for stuff like this

Reply:
I started with path.join, but found that it apparently had a side-effect of removing the "extra" slash in "http://localhost:3000" so that my absolute URL was treated as relative and I ended up with a URL of "http://localhost:3001/localhost:3000/whatever".
This code needs to handle both file paths and URLs, and I wasn't sure offhand of a single library to handle both.  This seemed the simplest thing since the append-with-slash is the functionality I really need here.  If you know a good option without the issue I discovered, I'm open to it.